clang/lib/Frontend/CASDependencyCollector.cpp

```diff
@@ -53,6 +53,16 @@ void CASDependencyCollector::finishedMainFile(DiagnosticsEngine &Diags) {
   std::vector<ObjectRef> Refs;
   Refs.reserve(Files.size());
   for (StringRef File : Files) {
+    // Exclude recording the filenames from DependencyOutputOptions::ExtraDeps.
+    // These are kept out of the compilation cache key since they can be
+    // passed-in and added at the point where the dependency file is generated,
+    // without needing to affect the cached compilation.
+    if (std::find_if(ExtraDeps.begin(), ExtraDeps.end(),
```

Comment:
This will not be correct if you have a `-fdepfile-entry=` that is also as a normal dep.  How about we have a `DependencyOutputOptions  dropExtraDeps(const DependencyOutputOptions &)` function and then change construction to
```
: DependencyFileGenerator(dropExtraDeps(Opts), llvm::vfs::makeNullOutputBackend())
```

That way we never add them in the first place.

Reply:
You mean `DependencyFileGenerator` doesn't see them but we add them back in from `CASDependencyCollector`, e.g. add them from `CASDependencyCollector::finishedMainFile()`, after the list of filenames is recorded in the CAS?

Reply:
Maybe add them in `CASDependencyCollector::replay` right before the cas filenames? Then it even matches the order it would be normally where extra deps are first.

Reply:
Isn't `CASDependencyCollector::replay` only called when there is a cache hit? And for a cache miss they'd need to be added from `finishedMainFile()`?

Reply:
A normal DependencyFileGenerator is created during compiler instance setup, which is what writes the file for a cache miss.  The CASDependencyCollector doesn't write any files except during replay.  Having the CASDependencyCollector do both the cas and the file writing might be fine, but we'd need to modify the compiler invocation so the normal DependencyFileGenerator doesn't get created.  Or we could disable the normal DependencyFileGenerator and also always replay the CASDependencyCollector even for a miss.

Reply:
> A normal DependencyFileGenerator is created during compiler instance setup, which is what writes the file for a cache miss.

Ah ok, I was under the impression that `CASDependencyCollector` takes over for misses as well. Applying the extra deps from `CASDependencyCollector::replay` seems ideal.

Reply:
See updated commit. I'm also testing for the "`-fdepfile-entry=` that is also as a normal dep" case.